File: experiments/whenwords-prompt-python/whenwords.py
"""
Whenwords: Human-friendly date and time formatting.

All functions are pure -- no side effects, no system clock access.
All dates are in UTC. Zero external dependencies.
"""

import math
import re
from datetime import datetime, timezone


def time_ago(timestamp: int, now: int) -> str:
    """
    Given a timestamp and a reference time, return a relative time string.
    Past: "3 hours ago". Future: "in 3 hours". Within ~45s: "just now".

    Uses natural-feeling thresholds with rounding.
    """
    diff = now - timestamp  # positive = past, negative = future
    abs_diff = abs(diff)

    # Thresholds (in seconds) and corresponding unit labels
    # (max_seconds_for_unit, unit_singular, unit_plural, divisor)
    thresholds = [
        (45, None, None, 1),                        # "just now"
        (90, "minute", "minutes", 60),               # 1 minute
        (45 * 60, "minute", "minutes", 60),          # N minutes (up to 45 min)
        (90 * 60, "hour", "hours", 3600),            # 1 hour
        (22 * 3600, "hour", "hours", 3600),          # N hours (up to 22 hours)
        (36 * 3600, "day", "days", 86400),           # 1 day
        (26 * 86400, "day", "days", 86400),          # N days (up to 26 days)
        (45 * 86400, "month", "months", 2592000),    # 1 month (~30 days)
        (320 * 86400, "month", "months", 2592000),   # N months
        (548 * 86400, "year", "years", 31536000),    # 1 year (~365 days)
        (float('inf'), "year", "years", 31536000),   # N years
    ]

    for max_secs, singular, plural, divisor in thresholds:
        if abs_diff < max_secs:
            if singular is None:
                return "just now"
            value = round(abs_diff / divisor)
            if value < 1:
                value = 1
            unit = singular if value == 1 else plural
            if diff > 0:
                return f"{value} {unit} ago"
            else:
                return f"in {value} {unit}"

    # Should not reach here, but just in case
    value = round(abs_diff / 31536000)
    if value < 1:
        value = 1
    unit = "year" if value == 1 else "years"
    if diff > 0:
        return f"{value} {unit} ago"
    else:
        return f"in {value} {unit}"


def duration(seconds: int, compact: bool = False, max_units: int = 2) -> str:
    """
    Format a number of seconds as a human-readable duration.

    Normal: "2 hours, 30 minutes"
    Compact: "2h 30m"

    Shows the `max_units` largest non-zero units (default 2).
    Rounds the last displayed unit rather than flooring.
    Raises ValueError on negative input.
    """
    if seconds < 0:
        raise ValueError("Duration cannot be negative")

    if seconds == 0:
        return "0s" if compact else "0 seconds"

    units = [
        ("day", "d", 86400),
        ("hour", "h", 3600),
        ("minute", "m", 60),
        ("second", "s", 1),
    ]

    # First pass: extract the largest units by flooring, then round the last one
    parts = []
    remaining = seconds

    for name, abbrev, divisor in units:
        if remaining >= divisor or (remaining > 0 and len(parts) > 0):
            value = remaining // divisor
            remaining -= value * divisor
            if value > 0 or len(parts) > 0:
                parts.append((name, abbrev, divisor, value))

    # If we have no parts (shouldn't happen for seconds > 0), fallback
    if not parts:
        return "0s" if compact else "0 seconds"

    # Now we need to select max_units and round the last one
    # Recalculate: pick the first max_units non-zero groups, rounding the last
    result_parts = []
    remaining = seconds

    for name, abbrev, divisor in units:
        value = remaining // divisor
        remaining -= value * divisor

        if value > 0 or len(result_parts) > 0:
            result_parts.append((name, abbrev, divisor, int(value), remaining))

    # Filter to only parts with non-zero values (initially)
    non_zero = [(n, a, d, v, r) for n, a, d, v, r in result_parts if v > 0]

    if not non_zero:
        # seconds was 0 but we already handled that
        return "0s" if compact else "0 seconds"

    # Take up to max_units
    selected = non_zero[:max_units]

    # Round the last selected unit based on the remaining seconds at that point
    if len(selected) > 0:
        last_name, last_abbrev, last_divisor, last_value, last_remaining = selected[-1]
        # Compute total remaining after this unit in original decomposition
        # We need to figure out how many seconds were left after the last unit
        # Recompute remaining after extracting selected units
        used = 0
        for n, a, d, v, r in selected:
            used += v * d
        leftover = seconds - used
        if leftover > 0 and leftover >= last_divisor / 2:
            last_value += 1
            selected[-1] = (last_name, last_abbrev, last_divisor, last_value, last_remaining)

    # Format
    if compact:
        return " ".join(f"{v}{a}" for _, a, _, v, _ in selected)
    else:
        formatted = []
        for name, _, _, value, _ in selected:
            unit = name if value == 1 else name + "s"
            formatted.append(f"{value} {unit}")
        return ", ".join(formatted)


def parse_duration(text: str) -> int:
    """
    Parse a human-written duration string back into seconds.

    Accepts compact ("2h30m"), verbose ("2 hours and 30 minutes"),
    colon notation ("2:30", "1:30:00"), decimals ("2.5 hours"),
    mixed with commas and "and".

    Case insensitive. Tolerates extra whitespace.
    Raises ValueError on empty strings, unrecognized input, negative values,
    and bare numbers without units.
    """
    if not isinstance(text, str):
        raise ValueError("Input must be a string")

    cleaned = text.strip()
    if not cleaned:
        raise ValueError("Empty duration string")

    # Check for negative values
    if cleaned.startswith('-'):
        raise ValueError("Negative durations are not allowed")

    # Try colon notation first: H:MM or H:MM:SS
    colon_match = re.fullmatch(r'(\d+):(\d{1,2})(?::(\d{1,2}))?', cleaned)
    if colon_match:
        hours = int(colon_match.group(1))
        minutes = int(colon_match.group(2))
        secs = int(colon_match.group(3)) if colon_match.group(3) else 0
        return hours * 3600 + minutes * 60 + secs

    # Normalize: lowercase, strip extra whitespace
    normalized = cleaned.lower()
    # Remove "and", commas
    normalized = re.sub(r'\band\b', ' ', normalized)
    normalized = normalized.replace(',', ' ')
    # Collapse whitespace
    normalized = re.sub(r'\s+', ' ', normalized).strip()

    # Unit mapping
    unit_map = {
        's': 1, 'sec': 1, 'secs': 1, 'second': 1, 'seconds': 1,
        'm': 60, 'min': 60, 'mins': 60, 'minute': 60, 'minutes': 60,
        'h': 3600, 'hr': 3600, 'hrs': 3600, 'hour': 3600, 'hours': 3600,
        'd': 86400, 'day': 86400, 'days': 86400,
        'w': 604800, 'wk': 604800, 'wks': 604800, 'week': 604800, 'weeks': 604800,
    }

    # Pattern to match number + unit pairs
    pattern = r'(\d+(?:\.\d+)?)\s*([a-z]+)'
    matches = re.findall(pattern, normalized)

    if not matches:
        # Check if it's just a bare number
        if re.fullmatch(r'\d+(?:\.\d+)?', normalized):
            raise ValueError("Bare number without units")
        raise ValueError(f"Unrecognized duration format: {text}")

    # Verify the entire string is consumed by our matches (plus separators)
    # Rebuild what we matched and compare
    rebuilt = normalized
    for num_str, unit_str in matches:
        rebuilt = rebuilt.replace(num_str + unit_str, '', 1)
        rebuilt = rebuilt.replace(num_str + ' ' + unit_str, '', 1)
    rebuilt = re.sub(r'[\s,]+', '', rebuilt)
    if rebuilt:
        raise ValueError(f"Unrecognized duration format: {text}")

    total = 0.0
    for num_str, unit_str in matches:
        if unit_str not in unit_map:
            raise ValueError(f"Unrecognized unit: {unit_str}")
        value = float(num_str)
        if value < 0:
            raise ValueError("Negative durations are not allowed")
        total += value * unit_map[unit_str]

    return round(total)


def human_date(timestamp: int, now: int) -> str:
    """
    Return a contextual date string.

    Today/yesterday/tomorrow use those words. Within ~a week: "Last Wednesday"
    or "This Friday". Same year: "March 5". Different year: "March 5, 2023".

    All date comparison based on calendar days in UTC.
    """
    dt = datetime.fromtimestamp(timestamp, tz=timezone.utc)
    now_dt = datetime.fromtimestamp(now, tz=timezone.utc)

    # Get calendar dates
    date_val = dt.date()
    now_date = now_dt.date()

    day_diff = (date_val - now_date).days  # positive = future, negative = past

    day_name = dt.strftime("%A")  # e.g., "Wednesday"
    month_name = dt.strftime("%B")  # e.g., "March"
    day_num = dt.day
    year = dt.year

    if day_diff == 0:
        return "Today"
    elif day_diff == -1:
        return "Yesterday"
    elif day_diff == 1:
        return "Tomorrow"
    elif -7 <= day_diff < -1:
        return f"Last {day_name}"
    elif 2 <= day_diff <= 7:
        return f"This {day_name}"
    elif dt.year == now_dt.year:
        return f"{month_name} {day_num}"
    else:
        return f"{month_name} {day_num}, {year}"


def date_range(start: int, end: int) -> str:
    """
    Format two timestamps as a smart date range string.

    Same day: single date. Same month: "January 15-22, 2024".
    Same year: "January 15 - February 15, 2024".
    Different years: "December 28, 2023 - January 15, 2024".

    Uses en-dash. Auto-swaps if start > end. All dates in UTC.
    """
    if start > end:
        start, end = end, start

    start_dt = datetime.fromtimestamp(start, tz=timezone.utc)
    end_dt = datetime.fromtimestamp(end, tz=timezone.utc)

    start_date = start_dt.date()
    end_date = end_dt.date()

    s_month = start_dt.strftime("%B")
    e_month = end_dt.strftime("%B")
    s_day = start_dt.day
    e_day = end_dt.day
    s_year = start_dt.year
    e_year = end_dt.year

    EN_DASH = "\u2013"

    if start_date == end_date:
        # Same day
        return f"{s_month} {s_day}, {s_year}"
    elif s_year == e_year and s_month == e_month:
        # Same month and year
        return f"{s_month} {s_day}{EN_DASH}{e_day}, {s_year}"
    elif s_year == e_year:
        # Same year, different months
        return f"{s_month} {s_day} {EN_DASH} {e_month} {e_day}, {s_year}"
    else:
        # Different years
        return f"{s_month} {s_day}, {s_year} {EN_DASH} {e_month} {e_day}, {e_year}"

File: research/experiments/whenwords-spec-go/whenwords.go
// Package whenwords provides pure functions for human-friendly date and time formatting.
// All functions accept explicit timestamps and never access the system clock.
package whenwords

import (
	"errors"
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// TimeAgo converts a Unix timestamp to a relative time string like "3 hours ago" or "in 2 days".
func TimeAgo(timestamp, reference int64) string {
	diff := reference - timestamp
	future := diff < 0
	seconds := diff
	if seconds < 0 {
		seconds = -seconds
	}

	var value int64
	var unit string

	switch {
	case seconds <= 44:
		return "just now"
	case seconds <= 89:
		value = 1
		unit = "minute"
	case seconds <= 2640:
		value = int64(math.Round(float64(seconds) / 60))
		unit = "minute"
	case seconds <= 5340:
		value = 1
		unit = "hour"
	case seconds <= 75600:
		value = int64(math.Round(float64(seconds) / 3600))
		unit = "hour"
	case seconds <= 126000:
		value = 1
		unit = "day"
	case seconds <= 2160000:
		value = int64(math.Round(float64(seconds) / 86400))
		unit = "day"
	case seconds <= 3888000:
		value = 1
		unit = "month"
	case seconds <= 27561600:
		value = int64(math.Round(float64(seconds) / 2592000))
		unit = "month"
	case seconds <= 47260800:
		value = 1
		unit = "year"
	default:
		value = int64(math.Round(float64(seconds) / 31536000))
		unit = "year"
	}

	// Pluralize
	if value != 1 {
		unit += "s"
	}

	if future {
		return fmt.Sprintf("in %d %s", value, unit)
	}
	return fmt.Sprintf("%d %s ago", value, unit)
}

// Duration formats a number of seconds as a human-readable duration string.
// compact uses abbreviated units ("2h 30m" vs "2 hours, 30 minutes").
// maxUnits controls the maximum number of units to display (default-like: 2).
// Panics on negative seconds.
func Duration(seconds int, compact bool, maxUnits int) string {
	if seconds < 0 {
		panic("duration seconds must be non-negative")
	}

	type unitDef struct {
		secs    int
		verbose string
		abbrev  string
	}

	units := []unitDef{
		{31536000, "year", "y"},
		{2592000, "month", "mo"},
		{86400, "day", "d"},
		{3600, "hour", "h"},
		{60, "minute", "m"},
		{1, "second", "s"},
	}

	if seconds == 0 {
		if compact {
			return "0s"
		}
		return "0 seconds"
	}

	// Decompose into units
	type part struct {
		value int
		unit  unitDef
	}
	var parts []part
	remaining := seconds
	for _, u := range units {
		if remaining >= u.secs {
			v := remaining / u.secs
			remaining = remaining % u.secs
			parts = append(parts, part{v, u})
		}
	}

	// Apply maxUnits with rounding on the last displayed unit
	if len(parts) > maxUnits {
		// Round up the last kept unit based on the remainder
		lastIdx := maxUnits - 1
		lastUnit := parts[lastIdx]

		// Calculate the total remaining seconds after the last kept unit
		remainderSecs := 0
		for i := maxUnits; i < len(parts); i++ {
			remainderSecs += parts[i].value * parts[i].unit.secs
		}

		// Round: if remainder >= half of the last unit's seconds, round up
		if remainderSecs*2 >= lastUnit.unit.secs {
			parts[lastIdx].value++
		}

		parts = parts[:maxUnits]
	}

	// Build output
	var strs []string
	for _, p := range parts {
		if compact {
			strs = append(strs, fmt.Sprintf("%d%s", p.value, p.unit.abbrev))
		} else {
			name := p.unit.verbose
			if p.value != 1 {
				name += "s"
			}
			strs = append(strs, fmt.Sprintf("%d %s", p.value, name))
		}
	}

	if compact {
		return strings.Join(strs, " ")
	}
	return strings.Join(strs, ", ")
}

var (
	errEmpty       = errors.New("empty duration string")
	errUnrecognized = errors.New("unrecognized duration format")
	errNegative    = errors.New("negative duration")
	errBareNumber  = errors.New("bare number without units")
	errUnknownUnit = errors.New("unknown unit")
)

// unitSeconds maps unit aliases (lowercase) to their value in seconds.
var unitSeconds = map[string]float64{
	"y": 31536000, "yr": 31536000, "yrs": 31536000, "year": 31536000, "years": 31536000,
	"mo": 2592000, "month": 2592000, "months": 2592000,
	"w": 604800, "wk": 604800, "wks": 604800, "week": 604800, "weeks": 604800,
	"d": 86400, "day": 86400, "days": 86400,
	"h": 3600, "hr": 3600, "hrs": 3600, "hour": 3600, "hours": 3600,
	"m": 60, "min": 60, "mins": 60, "minute": 60, "minutes": 60,
	"s": 1, "sec": 1, "secs": 1, "second": 1, "seconds": 1,
}

// Regex for matching number+unit pairs
var pairRegex = regexp.MustCompile(`(\d+(?:\.\d+)?)\s*([a-zA-Z]+)`)

// Regex for colon format
var colonRegex = regexp.MustCompile(`^(\d+):(\d{2})(?::(\d{2}))?$`)

// ParseDuration parses a human-written duration string into total seconds.
func ParseDuration(input string) (int, error) {
	s := strings.TrimSpace(input)
	if s == "" {
		return 0, errEmpty
	}

	// Check for negative
	if strings.HasPrefix(s, "-") {
		return 0, errNegative
	}

	// Try colon format first
	if m := colonRegex.FindStringSubmatch(s); m != nil {
		hours, _ := strconv.Atoi(m[1])
		minutes, _ := strconv.Atoi(m[2])
		total := hours*3600 + minutes*60
		if m[3] != "" {
			secs, _ := strconv.Atoi(m[3])
			total += secs
		}
		return total, nil
	}

	// Strip "and", commas for normalization
	normalized := strings.ReplaceAll(s, ",", " ")
	normalized = strings.ReplaceAll(normalized, " and ", " ")

	// Find all number+unit pairs
	matches := pairRegex.FindAllStringSubmatch(normalized, -1)
	if len(matches) == 0 {
		// Check if it's just a bare number
		trimmed := strings.TrimSpace(normalized)
		if _, err := strconv.ParseFloat(trimmed, 64); err == nil {
			return 0, errBareNumber
		}
		return 0, errUnrecognized
	}

	// Verify that the entire input is consumed by the pairs (plus allowed separators)
	// Build a version of normalized with all pairs removed
	remainder := normalized
	for _, m := range matches {
		remainder = strings.Replace(remainder, m[0], "", 1)
	}
	remainder = strings.TrimSpace(remainder)
	// Allow only whitespace, commas, "and" in remainder
	remainder = strings.ReplaceAll(remainder, "and", "")
	remainder = strings.ReplaceAll(remainder, ",", "")
	remainder = strings.TrimSpace(remainder)
	if remainder != "" {
		return 0, errUnrecognized
	}

	total := 0.0
	for _, m := range matches {
		numStr := m[1]
		unitStr := strings.ToLower(m[2])

		num, err := strconv.ParseFloat(numStr, 64)
		if err != nil {
			return 0, errUnrecognized
		}

		secs, ok := unitSeconds[unitStr]
		if !ok {
			return 0, errUnknownUnit
		}

		total += num * secs
	}

	return int(math.Round(total)), nil
}

// HumanDate returns a contextual date string based on proximity.
func HumanDate(timestamp, reference int64) string {
	ts := time.Unix(timestamp, 0).UTC()
	ref := time.Unix(reference, 0).UTC()

	// Truncate to dates
	tsDate := time.Date(ts.Year(), ts.Month(), ts.Day(), 0, 0, 0, 0, time.UTC)
	refDate := time.Date(ref.Year(), ref.Month(), ref.Day(), 0, 0, 0, 0, time.UTC)

	dayDiff := int(tsDate.Sub(refDate).Hours() / 24)

	switch {
	case dayDiff == 0:
		return "Today"
	case dayDiff == -1:
		return "Yesterday"
	case dayDiff == 1:
		return "Tomorrow"
	case dayDiff >= -6 && dayDiff <= -2:
		return "Last " + ts.Weekday().String()
	case dayDiff >= 2 && dayDiff <= 6:
		return "This " + ts.Weekday().String()
	case ts.Year() == ref.Year():
		return fmt.Sprintf("%s %d", ts.Month().String(), ts.Day())
	default:
		return fmt.Sprintf("%s %d, %d", ts.Month().String(), ts.Day(), ts.Year())
	}
}

// DateRange formats two timestamps as a smart date range.
func DateRange(start, end int64) string {
	if start > end {
		start, end = end, start
	}

	s := time.Unix(start, 0).UTC()
	e := time.Unix(end, 0).UTC()

	enDash := "\u2013"

	switch {
	case s.Year() == e.Year() && s.Month() == e.Month() && s.Day() == e.Day():
		// Same day
		return fmt.Sprintf("%s %d, %d", s.Month().String(), s.Day(), s.Year())
	case s.Year() == e.Year() && s.Month() == e.Month():
		// Same month
		return fmt.Sprintf("%s %d%s%d, %d", s.Month().String(), s.Day(), enDash, e.Day(), s.Year())
	case s.Year() == e.Year():
		// Same year, different month
		return fmt.Sprintf("%s %d %s %s %d, %d", s.Month().String(), s.Day(), enDash, e.Month().String(), e.Day(), s.Year())
	default:
		// Different years
		return fmt.Sprintf("%s %d, %d %s %s %d, %d", s.Month().String(), s.Day(), s.Year(), enDash, e.Month().String(), e.Day(), e.Year())
	}
}
